Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1338 @@
+
+### Parsing expressions
+
+Finally, let's move on to the last kind of production we want to parse: expressions.
+
+For now, we can parse three simple expression types: integers, variables, and strings.
+Expressions are often more complicated than this, as they may have function applications, list literals, local variable declarations, conditionals, etc.
+
+To begin, let's enrich our `Exp` AST with the remainder of the nodes:
+
+```haskell
+data Operator a
+  = Plus a
+  | Minus a
+  | Times a
+  | Divide a
+  | Eq a
+  | Neq a
+  | Lt a
+  | Le a
+  | Gt a
+  | Ge a
+  | And a
+  | Or a
+  deriving (Foldable, Show)
+
+data Exp a
+  = EInt a Integer
+  | EVar a (Name a)
+  | EString a ByteString
+  | EUnit a
+  | EList a [Exp a]
+  | EPar a (Exp a)
+  | EApp a (Exp a) (Exp a)
+  | EIfThen a (Exp a) (Exp a)
+  | EIfThenElse a (Exp a) (Exp a) (Exp a)
+  | ENeg a (Exp a)
+  | EBinOp a (Exp a) (Operator a) (Exp a)
+  | EOp a (Operator a)
+  | ELetIn a (Dec a) (Exp a)
+  deriving (Foldable, Show)
+```
+
+A few words and an example about each production:
+* `EInt`: An integer. `42`.
+* `EVar`: A variable. `my_variable`.
+* `EString`: A string. `"this is a string"`.
+* `EUnit`: The unit literal, with type `()`. `()`.
+* `EList`: A list literal. `[1, my_var, if true then 1 else 0`].
+* `EPar`: An expression within parentheses. `(1 + foo bar)`.
+* `EApp`: The application of an expression in a function. `my_func 1 "foo"`.
+* `EIfThen`: A conditional producing a unit. `if true then print "Hello"`.
+* `EIfThenElse`: A conditional. `if true then 1 else 2`.
+* `ENeg`: The negation of an expression. `-(x + 5)`.
+* `EBinOp`: The application of a binary operator to two expressions. `2 * 3 - 5`.
+* `EOp`: A binary operator without operands. `(+) x 1`.
+* `ELetIn`: Locally declares a variable. `let x = 5 in my_function x`.
+
+We've previously defined a few primitives, namely `integer`, `name`, and `string`.
+We may now add a few more simple productions.
+Just so we're on the same page, I'll repeat the productions of `exp` that we already had.
+
+```happy
+exp :: { Exp L.Range }
+  : integer                  { unTok $1 (\range (L.Integer int) -> EInt range int) }
+  | name                     { EVar (info $1) $1 }
+  | string                   { unTok $1 (\range (L.String string) -> EString range string) }
+  | '(' ')'                  { EUnit (L.rtRange $1 <-> L.rtRange $2) }
+  | '[' sepBy(exp, ',') ']'  { EList (L.rtRange $1 <-> L.rtRange $3) $2 }
+  | '(' exp ')'              { EPar (L.rtRange $1 <-> L.rtRange $3) $2 }
+```
+
+The only thing new here is the introduction of a new `sepBy` function, which should separate various productions (`exp`) by a separator (`,`), just like a list in Haskell.
+The definition of `sepBy` is provided below.
+
+```happy
+sepBy_rev(p, sep)
+  :                     { [] }
+  | sepBy(p, sep) sep p { $3 : $1 }
```

Comment:
Agh! Seems like I did the same mistake in `many` as well. :/

Fixing it actually caused 1 shift/reduce conflict in the solution of the exercises that can be resolved by using `many_rev` directly instead of `many`. Man, I wish Happy had an `%inline` directive...